### ISSUE TITLE
🧪 Testing Documentation and Github Action

### ISSUE THREAD

**Issue by jayehernandez** [jayehernandez/letra-extension#125]
This issue is created to add documentation and guidelines for testing with Jest.

References @tbarland77's Unit Testing Issue #79

To do for this includes:
- [ ] Make guideline for creating and running tests
- [ ] Add Jest Github Action which should run in every PR

**Comment by jayehernandez** [comment]
Hoping to finish everything in this issue within the week, stay tuned 👀

**Comment by tbarland77** [comment]
Glad to see some movement on the testing documentation! Let me know if there's anything I could help with. :thumbsup:

**Comment by jayehernandez** [comment]
Thanks @tbarland77. Trying to study it now but I can't seem to run the test command (`npm run test`). It displays this error:

```
> jest

 FAIL  test/example.test.js
  ● Test suite failed to run

    TypeError: (0 , _vm(...).compileFunction) is not a function

      at Runtime._execModule (node_modules/jest-runtime/build/index.js:1166:56)


=============================== Coverage summary ===============================
Statements   : 0% ( 0/58 )
Branches     : 0% ( 0/4 )
Functions    : 0% ( 0/9 )
Lines        : 0% ( 0/58 )
================================================================================
Test Suites: 1 failed, 1 total
Tests:       0 total
Snapshots:   0 total
Time:        1.3 s
Ran all test suites.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! letra@1.0.0 test: `jest`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the letra@1.0.0 test script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
``` 

Just want to ask if a similar error appeared when you were installing these? Tried it with both server and client and the same issue appears.

**Comment by tbarland77** [comment]
Hey, @jayehernandez I can't say I've seen that error. I would have called it out in my earlier PR if I was seeing that. I just tried it again on my end in the client:

![image](https://user-images.githubusercontent.com/15313520/83694626-61bb6480-a5c6-11ea-85e3-0680b3ea7206.png)


It appears to still be working for me minus the deprecation warning. I'll try deleting my node modules and do a fresh npm install to see if I can reproduce the error. I'll report back if I find anything.

**Comment by tbarland77** [comment]
Well, I tried various things but did have much luck. I was not able to reproduce the error even doing a clean npm install. There are only two things I can think of here:

1. Something weird happened with your node modules and maybe doing a fresh npm install will resolve the problem

2. Somehow the jest config is incorrect although if that's the case it makes no sense to me why it would work only for me. Anyway, I did not find anything too promising online but maybe you could give this a try?:

https://stackoverflow.com/questions/61973433/jest-encountered-an-unexpected-token-when-testing-a-vue-single-file-component 

Their transformations are a bit different from the ones I added following the guide I used but I tried using their changes and things still worked for me so maybe it's worth a try? That wouldn't really explain why you're also seeing the same error in the server though...

**Comment by jayehernandez** [comment]
Got it. Yeah, I assumed that it's really only happening with my machine, but just wanted to be sure. I tried doing a fresh `npm install` after deleting my `node_modules` folder but that didn't seem to work.

However, after some searching, I found a similar issue over a community in [Gitter](https://gitter.im/qawolf/community?at=5eb53d5d31a6d25d7cb87cfd) which suggested an upgrade of `node`. Upgrade from `v10.4.0` to `v12.18.0` (latest stable version as of writing) worked for me.

<img width="687" alt="Screen Shot 2020-06-04 at 5 27 56 PM" src="https://user-images.githubusercontent.com/13959651/83739906-c17a4400-a688-11ea-9e36-4cfc2d139199.png">

Thanks for helping with this! Will continue on with documenting, and I'll be sure to note this in the Troubleshooting section.

**Comment by tbarland77** [comment]
Interesting. I had not even considered that we might just have different node versions. :sweat_smile: That makes a lot of sense though especially since I had installed the latest version of Jest for both the client and server. Looking at my version of node I am on version 12 as well. Anyway, glad to hear you were able to resolve the problem and document it as well!

**Comment by jayehernandez** [comment]
I've just merged #130, thank you for the help @tbarland77. Please let me know if I missed anything.

**Comment by tbarland77** [comment]
Everything sounded pretty solid to me. We should probably just remove my example test in the server too once more tests are added. I only really added that to verify that jest was working correctly. Maybe I'll work on adding a few tests to both the client and server over the weekend so we can have some more useful examples.
